### ISSUE TITLE
connectToHEPBackend - timeouts?

### ISSUE THREAD

**Issue by solarmon** [sipcapture/HFP#3]
Hi,

Looking at the code for the  connectToHEPBackend function in https://github.com/sipcapture/HFP/blob/main/HFP.go I see that it uses the tls.Dial() and net.Dial() functions to connect to the HEP backend.

Both of these 'Dial' functions don't seem to have a timeout - as per:

https://pkg.go.dev/crypto/tls#Dial
https://pkg.go.dev/net#Dial

Is it correct to assume that since there is no timeout, then for any network/service issues that does not generate a failure response, the function will never time out? If it doesn't timeout then it will never generate an error and connectionStatus.Set(0) will not be executed to set the hfp_connection_status_out metric to 0?

Thus, the hfp_connection_status_out metric cannot be used to indicate the status of hfp in all network/service failure scenarios.

There is a timeout version for net.Dial:

https://pkg.go.dev/net#DialTimeout

However, there does not seem to be an equivalent for tls.Dial. But, here is a suggestion that 'DialContext' could potentially be used:

https://stackoverflow.com/questions/69353996/how-can-i-add-a-10-second-timeout-with-tls-dial-there-is-no-tls-dialtimeout-t

Could timeouts be implemented in connectToHEPBackend() to handle network/service timeouts appropriately?

**Comment by adubovikov** [comment]
Hi @solarmon ,

I am not sure if the timeout really necessary need it here, because it can be only work in case name resolution (DNS resolv timeout) in another cases the Dial will be stopped by kernel once the destinaion host is unreachable.

But yes, we will put this feature into our todo list.

Thank you

**Comment by solarmon** [comment]
@adubovikov 

Can you clarify what is meant by "Dial will be stopped by kernel once the destinaion host is unreachable".

I have used several methods to simulate (a network failure) on the host that HFP is running on:

* Routing the HEP backend IPs to a blackhole route
* Using firewall direct rules to DROP packets going to the HEP backend IPs.
* Using firewall direct rules to REJECT packets going to the HEP backend IPs.

For each of these methods, HFP does not seem to detect that failure, and the HFP logs even shows that it has sent successfully - "Sending HEP OUT successful without filters" - but the HEP packet never gets to the HEP backend.

**Comment by adubovikov** [comment]
@solarmon as I said, these params are in your kernel. If you set the drop packets, you can tune these params in your sysctl config (tcp,fin,ack, keepalive) . I.e. like here:

https://www.baeldung.com/linux/tcp-timeout
and more info here:

https://sysctl-explorer.net/net/ipv4/ (tcp_*)

Also, HFP has now SetKeepAlive timeout, you can also set to 1 second to detect if the socket is dead by third party manipulation, like in your case (firewall). 

Again, your kernel is responsible for all tcp settings, not an userspace app :)

P.S. and btw, Dial is only to establish connection, the rest is undercontrol of Write and Flush.

**Comment by adubovikov** [comment]
So, since the HFP supports KeepAlive and it works great, the only way to detect such annomalities is to activate -keepalive 1 (second interval) and it will solve all issues. 

For now I will close the ticket.

**Comment by solarmon** [comment]
@adubovikov

I have found that the reconnect only occurs when it next gets HEP packets. Thus, the status of the hfp_connection_status_out metric does not get updated (as to whether the session is up or down) until that happens. So this metric value cannot be used to determine the (potential) health of the connection to the HEP endpoint.

This might be fine in a single hfp server setup, with constant HEP packets. However in a clustered hfp setup, the standby node will not get HEP traffic, but you still want to know the health of its connection to the HEP endpoint.

The workaround I have implemented is to separately check for an established session TCP session to he HEP endpoint - for example to check if there is an existing connection to the HEP endpoint:

`netstat -tanp | grep hfp | grep ESTABLISHED | grep 9443 -c`

And if it does not exist then generate a local dummy/keepalive request to HFP to initiate the reconnect.

This dummy/keepalive request can just simply be a simple wget request - for example:

`wget --no-proxy 127.0.0.1:9060 --timeout=1 --tries=1 --background --quiet`

Or, using hepgen.js, a custom dummy HEP LOG packet can be generated that the HEP endpoint system could make use of - for example:

`./hepgen.js/hepgen.js -t tcp -s 127.0.0.1 -p 9060 -c ./conf/keepalive.js`

These dummy/keepalive requests then causes the hfp_connection_status_out metric to be updated appropriately.

I use this method with keepalived, which periodically runs the health check script to determine whether hfp is 'healthy' and whether to switch the virtual IP for the hfp cluster.
